File: MovieSerenity/MovieSerenity.Web/Modules/ServerTypes/MovieDB/PersonForm.ts
﻿import { StringEditor, DateEditor, EnumEditor, IntegerEditor, PrefixedContext, initFormType } from "@serenity-is/corelib";
import { PersonMovieGrid } from "../../MovieDB/Person/PersonMovieGrid";
import { Gender } from "../Modules/MovieDB.Person.Gender";

export interface PersonForm {
    FirstName: StringEditor;
    LastName: StringEditor;
    BirthDate: DateEditor;
    BirthPlace: StringEditor;
    Gender: EnumEditor;
    Height: IntegerEditor;
    MoviesGrid: PersonMovieGrid;
}

export class PersonForm extends PrefixedContext {
    static readonly formKey = 'MovieDb.Person';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!PersonForm.init)  {
            PersonForm.init = true;

            var w0 = StringEditor;
            var w1 = DateEditor;
            var w2 = EnumEditor;
            var w3 = IntegerEditor;
            var w4 = PersonMovieGrid;

            initFormType(PersonForm, [
                'FirstName', w0,
                'LastName', w0,
                'BirthDate', w1,
                'BirthPlace', w0,
                'Gender', w2,
                'Height', w3,
                'MoviesGrid', w4
            ]);
        }
    }
}

[Gender]; // referenced types

File: MovieSerenity/MovieSerenity.Web/Modules/MovieDB/Person/PersonForm.cs
using MovieSerenity.Modules.MovieDB.Person;
using Serenity.ComponentModel;
using System;

namespace MovieSerenity.MovieDb.Forms;

[FormScript("MovieDb.Person")]
[BasedOnRow(typeof(PersonRow), CheckNames = true)]
public class PersonForm
{
    public string FirstName { get; set; }
    public string LastName { get; set; }
    public DateTime BirthDate { get; set; }
    public string BirthPlace { get; set; }
    public Gender? Gender { get; set; }
    public int Height { get; set; }
}